addons/package_hierarchy/models/stock_quant_package.py

```diff
@@ -0,0 +1,223 @@
+        "stock.quant", string="All content", compute="_compute_children_quant_ids"
+    )
+    children_ids = fields.One2many("stock.quant.package", "parent_id", "Contained Packages")
+    depth = fields.Integer(compute="_compute_depth", store=True)
+
+    @api.depends("parent_id", "parent_id.top_parent_id")
+    def _compute_top_parent_id(self):
+        for pack in self:
+            parent = pack.parent_id
+            if not parent:
+                pack.top_parent_id = False
+            elif parent.top_parent_id:
+                pack.top_parent_id = parent.top_parent_id
+            else:
+                pack.top_parent_id = parent
+
+    @api.constrains("parent_id", "children_ids")
+    @api.onchange("parent_id", "children_ids")
+    def _constrain_depth(self):
+        max_package_depth = self.env.user.get_user_warehouse().u_max_package_depth
+        for pack in self:
+            top_parent = pack.top_parent_id or pack.parent_id
+            if top_parent.depth > max_package_depth:
+                raise ValidationError(_("Maximum package depth exceeded."))
+
+    @api.depends("children_ids", "children_ids.depth")
+    def _compute_depth(self):
+        """Is the max depth of any children"""
+        for pack in self:
+            children = pack.children_ids
+
+            if not children:
+                pack.depth = 1
+            else:
+                pack.depth = max(pack.children_ids.mapped("depth")) + 1
+
+    @api.constrains("parent_id")
+    def _check_parent_not_multi_location(self):
+        for parent_package in self.parent_id:
+            parent_package._check_not_multi_location()
+
+    @api.constrains("parent_id")
+    def _check_package_recursion(self):
+        if not self._check_recursion("parent_id"):
+            raise ValidationError("A package cannot be its own ancestor.")
+
+    def _check_not_multi_location(self):
+        for package in self:
+            locations = package.children_quant_ids.location_id
+            if len(locations) > 1:
+                raise ValidationError(
+                    _("Package cannot be in multiple " "locations:\n%s\n%s")
+                    % (package.name, ", ".join([l.name for l in locations]))
+                )
+
+    def _return_num_ancestors(self):
+        self.ensure_one()
+        return self.search_count([("id", "parent_of", self.id)]) - 1
+
+    def _return_ancestors(self):
+        self.ensure_one()
+        packages = self.search([("id", "parent_of", self.id)])
+        return packages - self
+
+    @api.depends(
+        "parent_id",
+        "children_ids",
+        "quant_ids.package_id",
+        "children_ids.quant_ids",
+        "children_ids.quant_ids.package_id",
+    )
+    def _compute_children_quant_ids(self):
+        for package in self:
+            if isinstance(package.id, models.NewId):
+                package.children_quant_ids = []
+            else:
+                package._check_not_multi_location()
```

Comment:
Do we want this check here?

If we keep the tree consistent when updating parent_id, shouldn't be enough?

Reply:
If this solely depended on children_ids I think this could be removed but I think we still want the check for the other situations covered in the depends statement.

Reply:
Shouldn't the constrain on parent_id pick this up anyway?